src/core/Refiner.ml

```diff
@@ -78,6 +79,25 @@ struct
     () |> RM.emit (RefineEnv.location env) @@ fun fmt () ->
     Format.fprintf fmt "Emitted hole:@,  @[<v>%a@]@." (S.pp_sequent ~lbl ctx) tp
 
+  let boundary_satisfied tm tp phi clo : bool m =
+    let* con = RM.lift_ev @@ Sem.eval tm in
+    let+ res = RM.trap @@ RM.abstract `Anon (D.TpPrf phi) @@ fun prf ->
+      RM.equate tp con @<< RM.lift_cmp @@ Sem.inst_tm_clo clo prf
+    in
+    Result.is_ok res
+
+
+  let print_boundary tm tp phi clo : unit m =
```

Comment:
maybe this function should be renamed, because it is printing a message about a hole